error.go

```diff
@@ -72,22 +65,242 @@ func collectMessages(data interface{}, messages []string) []string {
+	ValidationError
+}
+
+type JsonError struct {
+	LocalError
+}
+
+type JsonDeserializationError struct {
+	JsonError
+}
+
+type JsonSerializationError struct {
+	JsonError
+}
+
+type JsonNoDataError struct {
+	JsonError
+}
+
+type MissingParameterError struct {
+	LocalError
+}
+
+type MissingPropertyError struct {
+	LocalError
+}
+
+var MissingWebhookSignatureError = &LocalError{LibraryError{Message: MissingWebhookSignature}}
+var MismatchWebhookSignatureError = &LocalError{LibraryError{Message: MismatchWebhookSignature}}
+
+// API/HTTP error types
+
+// APIError represents an error that occurred while communicating with the EasyPost API.
+//
+// This is typically due to a specific HTTP status code, such as 4xx or 5xx.
+//
+// This is different from the Error class, which represents information about what triggered the failed request.
+//
+// The information from the top-level Error class is used to generate this error, and any sub-errors are stored in the Errors field.
+type APIError struct {
+	LibraryError
+	// Code is a machine-readable status of the problem encountered.
+	Code string
+	// StatusCode is the HTTP numerical status code of the response.
+	StatusCode int
+	// Errors may be provided if there are details about server-side issues that caused the API request to fail.
+	Errors []*Error `json:"errors,omitempty"`
 }
 
-// Error provides a pretty printed string of an error object based on present data.
+// Error provides a pretty printed string of an APIError object based on present data.
 func (e *APIError) Error() string {
 	if e.Message != "" {
 		if e.Code != "" {
-			return e.Code + " " + e.Message.(string)
+			return e.Code + " " + e.Message
 		}
-		return e.Message.(string)
+		return e.Message
 	}
 	if e.Code != "" {
 		return e.Code
 	}
-	return fmt.Sprintf("%d %s", e.StatusCode, e.Status)
+	return fmt.Sprintf("%d %s", e.StatusCode, e.Code)
+}
+
+type BadRequestError struct {
+	APIError
+}
+
+type ConnectionError struct {
+	APIError
+}
+
+type GatewayTimeoutError struct {
+	APIError
+}
+
+type InternalServerError struct {
+	APIError
+}
+
+type InvalidRequestError struct {
+	APIError
+}
+
+type MethodNotAllowedError struct {
+	APIError
+}
+
+type NotFoundError struct {
+	APIError
+}
+
+type PaymentError struct {
+	APIError
+}
+
+type ProxyError struct {
+	APIError
+}
+
+type RateLimitError struct {
+	APIError
+}
+
+type RedirectError struct {
+	APIError
+}
+
+type RetryError struct {
+	APIError
+}
+
+type ServiceUnavailableError struct {
+	APIError
+}
+
+type SSLError struct {
+	APIError
+}
+
+type TimeoutError struct {
+	APIError
+}
+
+type UnauthorizedError struct {
+	APIError
 }
 
-var EndOfPaginationError = errors.New("there are no more pages to retrieve")
+type UnknownHttpError struct {
+	APIError
+}
+
+// FromErrorResponse returns an APIError-based object based on the HTTP response.
+// Do not pass a non-error response to this function.
+func FromErrorResponse(response *http.Response) error {
```

Comment:
issue: can we rename this function? By the name alone I wouldn't know what this does. Verb - Action. Maybe something like `BuildErrorFromResponse`?

Reply:
Resolved